[DPE-4801] Block charm if plugin disable fails due to dependent objects

lucasgameiroborges opened the pull request:
## Issue

See https://github.com/canonical/postgresql-k8s-operator/issues/536

## Solution

If the charm detects objects that depend on plugins not enabled via config (most likely because the plugin was externally enabled) It will go into blocked state and warn the user. 

The user can then decide to either: Delete the objects and proceed with the plugin disabled (without updating the config) or update the juju config option of the plugin as enabled.

Local testing works, Will add unit/integration tests shortly.

marceloneppel reviewed (commented):
Thanks, Lucas! Nice UX!

Testing manually, I saw that the charm is not clearing the blocked status after the plugin is enabled.
```sql
CREATE EXTENSION IF NOT EXISTS pg_trgm WITH SCHEMA pg_catalog;
CREATE TABLE example(name VARCHAR);
CREATE INDEX example_idx ON example USING gin(name gin_trgm_ops);
```

```sh
juju config postgresql-k8s request_date_style="ISO, DMY"

# Wait for the charm do become blocked with the following message: Cannot disable plugins: Existing objects depend on it. See logs

juju config postgresql-k8s plugin_pg_trgm_enable=True

# Wait for the charm to become idle again. It's still blocked with the same message.
```

Also, will the charm become active again if the user deletes the index that uses the extension? I think we may need to have at least something that detects that.

lucasgameiroborges commented:
Thanks for the feedback @marceloneppel, I addressed the issues in the last commit, LMK if you still encounter them.

> Also, will the charm become active again if the user deletes the index that uses the extension? I think we may need to have at least something that detects that.

Just added a check inside `update_status`  that retries disabling the plugin. This was the fastest way to resolve the block in case the user decided to remove the dependent objects  themselves. Though I personally think waiting the ~5 minutes for the update_status call to resolve the block is a bit too much, but I don´t see any other way. Open to suggestions if you see a quicker way!

PS: Will add the missing tests once the behavior gets approved.

marceloneppel reviewed (approved):
LGTM!

The `update-status-hook` may take more than 5 minutes to fire in some environments because it's configurable.

As we don't have an observer like on the VM charm or any other mechanism right now, I think it's fine to keep the logic to set the active status back in the update status for now.

dragomirp commented on src/charm.py:
```diff
@@ -602,7 +603,7 @@ def _on_config_changed(self, event) -> None:
                 return
 
-    def enable_disable_extensions(self, database: str = None) -> None:
+    def enable_disable_extensions(self, database: str = None) -> None:  # noqa: C901
```
I know cyclomatic complexity is annoying, but we should either bump the limit or fit within it, not disable it on individual functions.

taurus-forever commented on src/charm.py:
```diff
@@ -1239,6 +1255,9 @@ def _on_update_status(self, _) -> None:
         if self._has_blocked_status or self._has_waiting_status:
+            # If charm was failing to disable plugin, try again (user may have removed the objects)
+            if self.unit.status.message == EXTENSION_OBJECT_MESSAGE:
+                self.enable_disable_extensions()
```
Do we want to call enable_disable_extensions() on every update-status?
This might slow-down PG with a lot of DBs.... which could be production.

IMHO, we can (should?) go into `error` state here and user have to call `juju resolve` after disabling of manually enabled extension.

lucasgameiroborges replied:
We only call it if the charm is in blocked status due to existing dependent objects. In normal workflow this shouldn't be an issue AFAICT, since 99.9% of the time the charm won't go inside the if statement. 

The problem I see with entering an `error` state (besides looking scary for the unsuspecting user) is that the charm doesn´t get the chance to auto-resolve: The most usual use-case (I think) is one where the user manually enables/uses an extension via SQL, and forgets to update juju config. They will then see the blocked status and logs and will go "oops lemme update this config my_plugin_enable=True" which will trigger the `config_changed` event and resolve itself.

This check inside juju status is for the (more unlikely) case where the user prefers to delete the objects and disable the extension. Here, if the charm sees that it was blocked due to dependent objects, it will retry removing the extension (in case the user deleted the offending objects).

That's why I would vouch for blocking the charm instead of going into error. @marceloneppel what do you think?

marceloneppel replied:
After hearing many times that we should avoid the error state, I prefer to use a blocked status here (it explicitly calls for action from the user instead of giving the impression that the charm is failing if it gets into an error state). As Lucas commented, it's scary to see the error status in a charm.

taurus-forever reviewed (approved):
LGTM, please add tests. I would go different approach, but up 2 Marcelo here.

P.S. Do we have understanding how expensive `enable_disable_extensions()`?
